p2-maven-plugin/src/main/java/org/eclipse/tycho/p2maven/tools/P2RepositoryDataManipulator.java

```diff
@@ -0,0 +1,136 @@
+/*******************************************************************************
+ * Copyright (c) 2025, 2025 Hannes Wellmann and others.
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Hannes Wellmann - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.tycho.p2maven.tools;
+
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiFunction;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+
+import org.apache.maven.model.Repository;
+import org.eclipse.equinox.internal.p2.repository.helpers.RepositoryHelper;
+import org.eclipse.equinox.p2.core.ProvisionException;
+import org.eclipse.equinox.p2.internal.repository.tools.XZCompressor;
+import org.eclipse.equinox.p2.repository.IRepository;
+import org.eclipse.tycho.p2maven.repository.P2RepositoryKind;
+import org.eclipse.tycho.p2maven.repository.P2RepositoryManager;
+
+@Named
+@Singleton
+public class P2RepositoryDataManipulator {
+
+	public static ModifiedRepositoryDescriptor createDescriptor(Repository repositoryLocation,
+			P2RepositoryKind repositoryKind, Path outputLocation, boolean compress, boolean xzCompress,
+			boolean keepNonXzIndexFiles) throws IllegalArgumentException {
+
+		if (outputLocation != null) {
+			try {
+				URI location = RepositoryHelper.localRepoURIHelper(new URI(repositoryLocation.getUrl()));
+				if (location.getScheme().equals("file") && Files.isSameFile(outputLocation, Path.of(location))) {
+					outputLocation = null; // Signal in-place edit
+				} else if (!Files.isDirectory(outputLocation)) {
+					Files.createDirectories(outputLocation);
+				} else {
+					try (Stream<Path> files = Files.list(outputLocation);) {
+						if (files.findAny().isPresent()) {
+							throw new IllegalStateException("Output location is not empty: " + outputLocation);
+						}
+					}
+				}
+				repositoryLocation.setUrl(location.toString()); // reset (potentially) normalized URL
+			} catch (IOException e) {
```

Comment:
I won't catch the exceptions here but rethrow them so the caller can possibly handle them.

Reply:
Makes sense.